### DISCUSSION TITLE
@click also triggers @click.outside when element is in shadow dom

### DISCUSSION THREAD

**Discussion by shashank-brightness** [alpinejs/alpine#4126] [1. Help]
i am using `@click` to open close some dropdown and then need to close it when someone clicks outside by using `@click.outside`. 
however because entire component is in shadow dom, it is not working as it should and `@click` also triggers `@click.outside`. 

I have created fiddle here to replicate the same behavior as my case.
https://jsfiddle.net/Shashank24390/omn3yr91/

If i transfer `@click.outside` to content which is opened/closed (where x-show is there) then it works because outside is not triggered till element visible as mentioned in [doc](https://alpinejs.dev/directives/on#outside) but i anyhow want to use `@click.outside` on same where `@click` itself is. 

For example when second time clicking on trigger element should not close it and if i need to do something else so for that purpose i want to detect whether **trigger element** or **outside of trigger element** is clicked.

but due to shadow dom, click also somehow trigger outside click and i am not sure why it is like that.

**Comment by ekwoka** [comment]
Do `@click.stop`

The core issue is that events that aren't composed, when they hit the shadow root barrier they are then turned into a new event that has the shadow root as the target.

This is part of how events work across the shadow Dom barrier.

So the easiest is that you just don't let your inside events go outside.

**Reply by shashank-brightness** [reply]
If I use `@click.stop` then clicking on toggle 2 is not closing toggle 1 content and so on because that is also stopped in that case!!

**Reply by ekwoka** [reply]
Well, you can dispatch your own event, or put the click outside and a click.stop on the panel itself.

A few ways.

**Reply by shashank-brightness** [reply]
If i need to detect click outside anywhere, is there any option available which can even work in shadow dom without having to separate out click on parent child and so on..? I need to be able to simply write `@click="something" @click.outside="something else"` and that on same element.

**Reply by ekwoka** [reply]
You can normally.

But when the event passes from the shadow Dom, the target is changed to the shadow root.

So if you need to have your own handling to filter the event by it not being from the current shadow root.

You have to tools.

I believe in you!!!

**Comment by SimoTod** [comment]
Because of the way the shadodom works, you need to add an additional check inside the callback: `@click.outside="if($event.composedPath()[0] != $el) {console.log('click outside triggered also with click!!!!');hide()}"`
I think it only works if mode=open but that seems to be the case for you.
